Introduce 10.13 xnu

toonn opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

By introducing the new XNU without disturbing the old version, we avoid causing a stdenv rebuild. We can still depend on this new XNU in a new bootstrap-tools and further source releases.

I'm doing this in the hope it'll enable smaller reviewable changes rather than needing to understand everything all at once.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

toonn committed:
apple-source-releases: Add 10.13.6 releases

The `applePackageMapping` function will allow us to keep expressions for
the newer releases separate from the originals where necessary, without
needing to change `generate-sdk-packages.sh`. If we ever want to support
multiple versions of source releases simultaneously we should consider
changing that to point to versioned subdirectories of all the packages.

Many of the source releases introduced here will need changes to the
expression to build successfully. These will be introduced one by one.

toonn committed:
darwin."10.13.6".xnu: Introduce XNU version for macOS 10.13.6

By introducing the new XNU without disturbing the old version, we avoid
causing a stdenv rebuild. We can still depend on this new XNU in a new
bootstrap-tools and further source releases.

I'm doing this in the hope it'll enable smaller reviewable changes
rather than needing to understand everything all at once.

toonn committed:
darwin."10.13.6".xnu: Drop outdated substitutes

toonn committed:
darwin."10.13.6".xnu: Raise SDK and HOST_OS versions

We need to raise the `HOST_OS_VERSION` to at least `10.12` according to
an `ld` warning. Since we're bumping anyway I don't see a good reason
not to bump these all the way to the source releases we'll be packaging.

toonn committed:
darwin."10.13.6".xnu: Drop unnecessary -f from rm

toonn committed:
darwin."10.13.6".xnu: Complete versions in fake availability.pl

These are the versions from `AvailabilityVersions/availability.pl` for
macOS 10.13.6. The only reason we create the shell script is to avoid a
dependency on Perl in the bootstrap.

We should consider packaging AvailabilityVersions so the script becomes
something that consumes `availability.pl`, that way we don't need to
keep the versions synchronized manually.

toonn committed:
darwin."10.13.6".xnu: Implement missing availability platform

The `bridgeos(x)` platform function isn't defined in Apple's XNU but it
is used in Security, so we implement it here.

toonn committed:
darwin."10.13.6": Override the scope

Releases with dependencies on other releases, e.g., libresolv depends on
Libinfo, mDNSResponder and configdHeaders, were getting those releases
from pkgs.darwin rather than pkgs.darwin."10.13.6". By making the scope
of callPackage overridable we can manipulate which packages are used to
supply dependencies.

toonn committed:
darwin."10.13.6".xnu: Adopt

paparodeo commented on pkgs/os-specific/darwin/apple-source-releases/xnu/10.13.6.nix:
```diff
@@ -0,0 +1,165 @@
+      --replace-fail "-Werror " ""
+
+    substituteInPlace SETUP/kextsymboltool/Makefile \
+      --replace-fail "-lstdc++" "-lc++ -lc++abi"
```
nit: `-lc++abi` is not needed.
```suggestion
      --replace-fail "-lstdc++" "-lc++"
```
tho i think that clang changes `-lstdc++` to `-lc++` too so the entire substituteInPlace might be unneeded as well as long as the link step is initiated by clang and not ld

toonn replied:
It still builds without the substitute but I can't find documentation on this. Do you know where to find some?

paparodeo replied:
i don't i just stumbled on it. it seems like it is related to the flag `-stdlib=xxx`. when `-stdlib=libstdc++` no translation is made. `-stdlib=libc++`  then `-lstdc++ -> -lc++`. and perhaps the default is `-stdlib=platform` and for darwin seems to be `libc++`.

toonn committed:
darwin."10.13.6": Only include working source releases

Some of the source releases build without changes to their expressions,
others require changes or the passing in of exceptional arguments. To
avoid evaluation of the entire "10.13.6" subset causing evaluation
errors we can include just the functional source releases.

toonn committed:
darwin."10.13.6".xnu: Drop redundant substitutes

Co-authored-by: paparodeo <170618376+paparodeo@users.noreply.github.com>

paparodeo reviewed (approved):
LGTM tho i don't have any domain knowledge here. seems to mostly parallel `default.nix`

toonn committed:
darwin."10.13.6.xnu: Use --replace-fail

toonn commented on pkgs/os-specific/darwin/apple-source-releases/xnu/10.13.6.nix:
```diff
@@ -0,0 +1,159 @@
+    substituteInPlace makedefs/MakeInc.top \
+      --replace-fail "MEMORY_SIZE := " 'MEMORY_SIZE := 1073741824 # '
+
+    substituteInPlace libsyscall/xcodescripts/mach_install_mig.sh \
```
I figure we don't actually know better than Apple with regard to the header that affects.

Noticed I forgot to drop the trailing backslash though. That's the only real change in the new force-push.

The 10.13.6.nix is supposed to be easily diffable with default.nix, without affecting the old source release builds yet. Switching over will be done all at once, probably renaming all the 10.13.6 -> default. Though depending on whether aarch64-darwin needs certain source releases to build it's possible some sort of split will be retained.

paparodeo replied:
ah-- i just looked at the source and see that

```sh
if [[ ( "$MACHINE_ARCH" =~ ^"arm64" ) ]]
then
    # MACHINE_ARCH *really* needs to be a 32-bit arch to generate vm_map_internal.h correctly, even if there are no 32-bit targets.
    MACHINE_ARCH="armv7"
fi
```
still exists -- i think that will create issues with cc-wrapper as it adds aarch64 specific flags which makes clang fail. i don't think the x64 build adds cpu specific flags so x64 works fine.

toonn replied:
Is the XNU source build even used on aarch64-darwin though? It shouldn't be.

And even so, the proper fix here would be to disable those flags for this invocation, "*really* needs to be a 32-bit arch" doesn't sound to me like "it'll work fine even with arm64."

paparodeo replied:
> Is the XNU source build even used on aarch64-darwin though? It shouldn't be.

the default.nix one gets built but beyond that i have no other knowledge.
 
> And even so, the proper fix here would be to disable those flags for this invocation, "_really_ needs to be a 32-bit arch" doesn't sound to me like "it'll work fine even with arm64."

i agree completely.
